{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment - WSI specific\n",
    "\n",
    "## Introduction\n",
    "\n",
    "In this tutorial, will then align whole-genome sequencing from a mouse zygote which was subject to CRISPR-induced mutagenesis. We will find the resulting engineered alleles, and track down some other alleles in this mouse.  \n",
    "\n",
    "## Authors\n",
    "This tutorial was written by [Vivek Iyer](https://www.sanger.ac.uk/people/directory/iyer-vivek) and [Thomas Keane](https://github.com/tk2). It was adapted for the WSI environment by Vivek Iyer\n",
    "\n",
    "## Tutorial sections\n",
    "This tutorial comprises the following sections:   \n",
    " 1. [Visualising the the reference genome](reference_genome_visualisation.ipynb)   \n",
    " 2. [Aligning paired FASTQ files with BWA](bwa_alignment.ipynb)   \n",
    " 3. [Converting a SAM file to a BAM file](sam_bam_conversion.ipynb)   \n",
    " 4. [Sorting and indexing the BAM file](sort_and_index_bam.ipynb)   \n",
    " 5. [Using Unix pipes to combine the commands together](piping_commands.ipynb)\n",
    " 6. [Marking PCR duplicates](pcr_duplicates.ipynb)\n",
    " 7. [Generating QC stats](qc_stats.ipynb)\n",
    " 8. [BAM visualisation](bam_visualisation.ipynb)\n",
    " \n",
    "\n",
    "## Running the commands from this tutorial\n",
    "The commands in this tutorial are meant to be run directly in a WSI (Sanger) server environment - in our case by logging into the server node vr-login. We have pre-prepared a working area for each user: /lustre/scratch115/teams/hgi//WSI_NGS_tutorial/read_alignment/Exercise2/<user_name> and all directories specified are underneath this directory.\n",
    "\n",
    "### Sanger user id, 'ssh gateway' and 'network' passwords\n",
    "We assume you have a Sanger User Id, which we will refer to as *_user_* and that you have used it to \n",
    "\n",
    "* Log into the sanger ssh gateway\n",
    "* Log into individual servers\n",
    "\n",
    "### Logging into the WSI server environment and running commands in the terminal\n",
    "To run these commands you first have to open a terminal window, log into the WSI compute environment and navigate to the working directory. This is similar to the \"Command Prompt\" window on MS Windows systems, which allows the user to type DOS commands to manage files.\n",
    "\n",
    "**Open a terminal window in your VM**\n",
    "\n",
    "**Log into the Sanger server environment by entering the following command at the prompt:**\n",
    "\n",
    "    $ ssh -L localhost:3128:wwwcache.sanger.ac.uk:3128 vvi@ssh.sanger.ac.uk\n",
    "\n",
    "_where 'vvi' should be replaced with your user name (e.g. \"kj6\" or \"ls7\" etc)_\n",
    "\n",
    "**You will receive a password prompt: enter your ssh gateway password.** \n",
    "\n",
    "    vvi@ssh.sanger.ac.uk's password: \n",
    "    Last login: Sun Aug 25 08:30:51 2019 from cpc91218-cmbg18\n",
    "                  Wellcome Sanger Institute\n",
    "                         SSH Gateway\n",
    "\n",
    "         ********************************************\n",
    "         * This system is for authorised users only *\n",
    "         ********************************************\n",
    "\n",
    "**After entering the ssh gateway, you will be prompted \"Where would you like to go today\"?  Enter the server \"vr-login\".**\n",
    "\n",
    "    Where would you like to go today (exit to logout)? \n",
    "\n",
    "        > vr-login\n",
    " \n",
    "**You will then be prompted for your standard network password:**\n",
    "\n",
    "    Connecting to vr-login.internal.sanger.ac.uk ...\n",
    "    vvi@vr-login.internal.sanger.ac.uk's password: \n",
    "    \n",
    "**On providing this, you will get access to a command prompt on vr-login**\n",
    "\n",
    "    Welcome to Ubuntu 12.04.2 LTS (GNU/Linux 3.2.0-105-generic x86_64)\n",
    "     * Documentation:  https://help.ubuntu.com/\n",
    "    This system is managed by CFEngine 3\n",
    "    Last login: Sun Aug 25 11:11:46 2019 from ssh.sanger.ac.uk\n",
    "    vvi@vr-2-2-02:~$ \n",
    "    \n",
    "**Set up your software environment to mirror NPG's environment**\n",
    "\n",
    "    vvi@vr-2-2-02:~$ . /software/npg/etc/profile.npg\n",
    "   \n",
    "This will allow you to access the same software (samtools, bcftools etc) that NPG uses\n",
    "\n",
    "**Navigate to your working area of the course, replacing 'vvi' with your user name**\n",
    "\n",
    "    vvi@vr-2-2-02:~$ cd /lustre/scratch115/teams/hgi/WSI_NGS_tutorial/read_alignment/Exercise2/vvi/\n",
    "    \n",
    "All following commands assume that you are in this directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Letâ€™s get started!\n",
    "\n",
    "To get started with the tutorial, head to the first section: [Visualising the the reference genome](reference_genome_visualisation.ipynb)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
